### PR TITLE
DOC: Added examples for correlation, num_obs_dm, num_obs_y functions.

### PR DESCRIPTION
**Pull request by tamri**
Reference issue
DOC: Add "Examples" to docstrings https://github.com/scipy/scipy/issues/7168

What does this implement/fix?
Added examples for correlation, num_obs_dm, num_obs_y functions.

Additional information
[skip actions] [skip cirrus]

### CODE REVIEW

**Comment by ilayn** [comment]
Hi @tamri Thanks a lot for your efforts to help us! Could you please add the `[skip actions] [skip cirrus]` identifiers to your commit messages? 

Unfortunately GitHub doesn't read the issue description and needs these identifiers at every commit to skip tests.

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -620,6 +620,23 @@ def correlation(u, v, w=None, centered=True):
     correlation : double
         The correlation distance between 1-D array `u` and `v`.
 
+    Examples
+    --------
+    Find the correlation between two arrays.
+
+    >>> from scipy.spatial.distance import correlation
+    >>> correlation([1, 0, 1],[1, 1, 0])
```
Comment (dschmitz89) [inline scipy/spatial/distance.py:628]:
```suggestion
    >>> correlation([1, 0, 1], [1, 1, 0])
```

File: scipy/spatial/distance.py [hunk 2]
```diff
@@ -620,6 +620,23 @@ def correlation(u, v, w=None, centered=True):
...
+    1.5
+
+    Using a weighting array, the correlation can be calculated as:
+
+    >>> correlation([1, 0, 1],[1, 1, 0], [0.9, 0.1, 0.1])
```
Comment (dschmitz89) [inline scipy/spatial/distance.py:633]:
```suggestion
    >>> correlation([1, 0, 1], [1, 1, 0], w=[0.9, 0.1, 0.1])
```

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Thanks for the PR @tamri . 

The linting issues are currently not caught by the linter infrastructure.

I do not really understand the purpose of `num_obs_dm` and `num_obs_y`, so no comment on those examples itself from my side. For the correlation distance, I find the examples helpful.

**Comment by tamri** [comment]
@dschmitz89 thx for checking the linting, will be careful about that next time.
Are you familiar with the linux_aarch64_test? It fails because "Monthly free compute limit exceeded!"

**Comment by dschmitz89** [comment]
> @dschmitz89 thx for checking the linting, will be careful about that next time.
> Are you familiar with the linux_aarch64_test? It fails because "Monthly free compute limit exceeded!"
> 

That is unrelated to this PR. CI providers only grant some amount of free computation resources per month. For Cirrus, we run out of them regularly unfortunately now.

**Comment by tamri** [comment]
Anything else needed for this PR to be merged?

**Review by dschmitz89 — COMMENTED** [review]
My comments were addressed but this still needs a look from a maintainer.

**Comment by j-bowhay** [comment]
The examples seem fine however I wondering if they could be adapted to provide more insight beyond just plugging in inputs? If the answer is the answer is no, not easily, that is also fine as this is still an improvement on no example

**Comment by dschmitz89** [comment]
There are actually two other (rather stale) PRs looking into the same functions: #18431 and #17785 . This one looks closest to me. When it is merged, the other PRs should be closed. 

Wondering if we should ask people to indicate in #7168 if they are working on a specific function to avoid duplicate work. That's how scikit learn  folks do it on a similar meta issue also: https://github.com/scikit-learn/scikit-learn/issues/27090 . However, I did not see duplicate work happening so often in scipy.

**Comment by tamri** [comment]
@j-bowhay If you could indicate a golden sample that shows what you mean by more insight I can try to follow it.
During my last PR, I tried to be thorough but ended up adding unnecessary redundancy.
Anyway, I'll wait until you confirm how you would like to move forward.

**Review by j-bowhay — APPROVED** [review]
Sorry this slipped under the radar. On second thought, these are quite simple functions so these examples are probably fine and can always be added to in the future if needed. For a more important function, a good narrative to the example is nice eg: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sobol_indices.html#scipy.stats.sobol_indices

File: scipy/spatial/distance.py [hunk 3]
```diff
@@ -620,6 +620,23 @@ def correlation(u, v, w=None, centered=True):
...
+    1.5
+
+    Using a weighting array, the correlation can be calculated as:
+
+    >>> correlation([1, 0, 1], [1, 1, 0], w=[0.9, 0.1, 0.1])
+    1.1
+
+    If centering is not needed, the correlation can be calculated as:
+
+    >>> correlation([1, 0, 1], [1, 1, 0], centered = False)
```
Comment (j-bowhay) [inline scipy/spatial/distance.py:638]:
```suggestion
    >>> correlation([1, 0, 1], [1, 1, 0], centered=False)
```

File: scipy/spatial/distance.py [hunk 4]
```diff
@@ -2601,6 +2618,15 @@ def num_obs_dm(d):
     num_obs_dm : int
         The number of observations in the redundant distance matrix.
 
+    Examples
+    --------
+    Find the number of original observations corresponding
+    to a square redundant distance matrix d.
+    
+    >>> from scipy.spatial.distance import num_obs_dm
+    >>> d = [[0, 100, 200],[100, 0, 150],[200, 150, 0]]
```
Comment (j-bowhay) [inline scipy/spatial/distance.py:2627]:
```suggestion
    >>> d = [[0, 100, 200], [100, 0, 150], [200, 150, 0]]
```

**Pull request merged**